basestar-storage-sql/src/main/java/io/basestar/storage/sql/SQLDialect.java

```diff
@@ -851,4 +856,43 @@ default SQLExpressionVisitor expressionResolver(final NamingStrategy namingStrat
         final InferenceContext inferenceContext = InferenceContext.from(schema);
         return new SQLExpressionVisitor(this, inferenceContext, columnResolver);
     }
+
+    default int createObjectLayer(final DSLContext context, final org.jooq.Table<?> table, final Field<String> idField, final String id, final Map<Field<?>, SelectField<?>> record) {
+
+        return context.insertInto(table)
+                .columns(record.keySet())
+                .select(DSL.select(record.values().toArray(new SelectFieldOrAsterisk[0])))
+                .execute();
+    }
+
+    default int updateObjectLayer(final DSLContext context, final org.jooq.Table<?> table, final Field<String> idField, final Field<Long> versionField, final String id, final Long version, final Map<Field<?>, SelectField<?>> record) {
+
+        Condition condition = idField.eq(id);
+        if (version != null) {
+            condition = condition.and(versionField.eq(version));
+        }
+
+        return context.update(table).set(record)
+                .where(condition).limit(DSL.inline(1)).execute();
+    }
+
+    default int createHistoryLayer(final DSLContext context, final org.jooq.Table<?> table, final Field<String> idField, final Field<Long> versionField, final String id, final Long version, final Map<Field<?>, SelectField<?>> record) {
+
+        context.deleteFrom(table).where(idField.eq(id).and(versionField.eq(version))).execute();
+
+        return context.insertInto(table)
+                .columns(record.keySet())
+                .select(DSL.select(record.values().toArray(new SelectFieldOrAsterisk[0])))
```

Comment:
I don't think `.keySet()` and `.values()` are guaranteed to be the same order in a `Map`, especially when used as `Set/Collection` here. Should it be converted to a list of pairs?

Reply:
In practise the order is deterministic (although unpredictable) given any instance of a 'standard' map, but I will change it anyway, and maybe order it explicitly, because there can def. be different orders in different JVMs.